File: pyshell/test_script/test_background.c
#include <unistd.h>
#include <stdio.h>
#include <sys/wait.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>


#define SIGNAL_STR(var, signal)	if (var == signal) signal_str = #signal;

void	interrupt_handler(int signum)
{
	char *signal_str = "";

	SIGNAL_STR(signum, SIGTSTP);
	SIGNAL_STR(signum, SIGCONT);
	printf("Father catched %s.\n", signal_str);
	signal(signum, SIG_DFL);
	kill(getpid(), signum);
}

void	show_tab(char **array)
{
	while (*array)
		printf("%s\n", *array++);
}

void	run_child(char **argv, char **environ)
{
	char *execname = argv[1];
	char **args_exec = argv + 1;

	signal (SIGINT, SIG_DFL);
	signal (SIGQUIT, SIG_DFL);
	signal (SIGTSTP, SIG_DFL);
	signal (SIGTTIN, SIG_DFL);
	signal (SIGTTOU, SIG_DFL);
	signal (SIGCHLD, SIG_DFL);
	setpgid(0, 0);
	tcsetpgrp(0, getpgrp());
	if (execve(execname, args_exec, environ) == -1)
		perror("execve");
}

void	error(char *str)
{
	printf("%s\n", str);
	exit(1);
}

char	*read_line(void)
{
	char buffer[250];
	int ret;

	ret = read(0, buffer, 249);
	buffer[ret] = 0;
	printf("%s", buffer);
	return (NULL);
}

void	init_shell(void)
{
	signal (SIGINT, SIG_IGN);
	signal (SIGQUIT, SIG_IGN);
	signal (SIGTSTP, SIG_IGN);
	signal (SIGTTIN, SIG_IGN);
	signal (SIGTTOU, SIG_IGN);
	signal (SIGCHLD, SIG_IGN);
}

 /*
 ** Fork a process a pull it in background by changing his process group id (pgid).
 ** Wait a while with the father, restore de children in the foreground process group (with tcsetgrp)
 ** and send him SIGCONT.
 */


int		main(int argc, char **argv, char **environ)
{
	pid_t pid;
	char *argv_prog[] = {"top",  NULL};
	char *ps_argv[] = {"ps", "-au", "simrossi", "-o", "user,command,pgid,gid,tpgid", NULL};
	int		status;
	int term_sig;
	int return_val;

	if (argc == 1)
		error("No args.");
	init_shell();
	if ((pid = fork()) == -1)
	{
		perror("fork");
		return (0);
	}
	if (pid == 0)
	{
		run_child(argv, environ);
	}
	else
	{
		setpgid(pid, 0);
		tcsetpgrp(0, pid);
		printf("Father starting wait.\n");
		waitpid(pid, &status, WUNTRACED);
		printf("Father end wait.\n");
		sleep(2);
		printf("end father sleep\n");
		tcsetpgrp(0, getpgrp());
		read_line();
		return_val = WEXITSTATUS(status);
		term_sig = WSTOPSIG(status);
		if (term_sig == SIGTSTP)
		{
			sleep(3);
			printf("Yolo\n");
		}
	}
	return (0);
}

File: pyshell/test/test_script/test_background.py
#!/usr/bin/env python3

import os, time, signal
import sys


def init_shell():
    signal.signal(signal.SIGINT, signal.SIG_IGN)
    signal.signal(signal.SIGQUIT, signal.SIG_IGN)
    signal.signal(signal.SIGTSTP, signal.SIG_IGN)
    signal.signal(signal.SIGTTIN, signal.SIG_IGN)
    signal.signal(signal.SIGTTOU, signal.SIG_IGN)
    signal.signal(signal.SIGCHLD, signal.SIG_IGN)


def run_child():

    execname = sys.argv[1]
    args_exec = sys.argv[1:]

    signal.signal(signal.SIGINT, signal.SIG_DFL);
    signal.signal(signal.SIGQUIT, signal.SIG_DFL);
    signal.signal(signal.SIGTSTP, signal.SIG_DFL);
    signal.signal(signal.SIGTTIN, signal.SIG_DFL);
    signal.signal(signal.SIGTTOU, signal.SIG_DFL);
    signal.signal(signal.SIGCHLD, signal.SIG_DFL);
    os.setpgid(0, 0)
    os.tcsetpgrp(0, os.getpgrp())
    os.execve(execname, args_exec, os.environ)


def error(msg):
    print(msg, file=sys.stderr)
    exit(1)

def read_line():
    print(os.read(0, 250).decode(), end='')



def main():
    pid = os.fork()
    if pid == 0:
        run_child()#args
    else:
        os.setpgid(pid, 0)
        os.tcsetpgrp(0, pid)
        print("Father starting wait.")
        status = os.waitpid(pid, os.WUNTRACED)
        print("Father end wait.")
        os.tcsetpgrp(0, os.getpgrp())
        if os.WIFSTOPPED(status[1]):
            wait_time = 2
            print("Relaunch if {} seconds.".format(wait_time))
            time.sleep(wait_time)
            os.tcsetpgrp(0, pid)
            os.kill(pid, signal.SIGCONT)
            status = os.waitpid(pid, os.WUNTRACED)
            os.tcsetpgrp(0, os.getpgrp())
            print("End Father.")
            exit(1)
        read_line()
        term_sig = os.WSTOPSIG(status[1])
        if (term_sig == signal.SIGTSTP):
            print("Yolo")

    pass

if __name__ == "__main__":
    if len(sys.argv) == 1:
        error("No args.")
    init_shell()
    main()
